### PR TITLE
Introduce OpenSearch integration

### LINKED ISSUE #1918: [PROPOSAL] Introduce `spring-data-opensearch` to support OpenSearch integration
Picking up the discussion started in https://github.com/spring-projects/spring-data-elasticsearch/issues/1770, at this moment it becomes clear that both projects are going to evolve in (slightly) incompatible fashion. Taking into account how widely popular Spring Data Elasticsearch is, it is highly likely that community would seek the first class support of the OpenSearch  integration as well. 

In scope of this issue, it would be great to hear the feedback on the proposal to introduce dedicated `spring-data-opensearch` project so to have OpenSearch supported out of the box, not depending on Elasticsearch. @sothawo if there is an interest and willing to accept the contribution, I would be more than happy to work on the pull request.

The OpenSearch project just published the Maven artifacts for client libraries [1], there should be no issues to rely on those for `spring-data-opensearch` and disconnect from Elasticsearch.

Thank you very much.

[1] https://discuss.opendistrocommunity.dev/t/maven-repository-artifacts-for-plugin-development/6406/7]

### PR DESCRIPTION
**Pull request by reta**
- [X] You have read the [Spring Data contribution guidelines](https://github.com/spring-projects/spring-data-build/blob/master/CONTRIBUTING.adoc).
- [X] **There is a ticket in the bug tracker for the project in our [issue tracker](https://github.
  com/spring-projects/spring-data-elasticsearch/issues)**. Add the issue number to the _Closes #issue-number_ line below
- [x] You use the code formatters provided [here](https://github.com/spring-projects/spring-data-build/tree/master/etc/ide) and have them applied to your changes. Don’t submit any formatting related changes.
- [x] You submit test cases (unit or integration tests) that back your changes.
- [x] You added yourself as author in the headers of the classes you touched. Amend the date range in the Apache license header if needed. For new types, add the license header (copy from another file and set the current year only).

Since https://github.com/spring-projects/spring-data-elasticsearch/pull/2196/files got merged, we have an opportunity to introduce clean OpenSearch integration (based on `HighLevelRestClient` at least) along with Elasticsearch one. This pull request draft aims to kick of the discussion if such a path forward makes sense for maintainers or not. Upon positive feedback, the remaining issues (test cases, documentation, ...) are going to be added right away.

@sothawo what do you think about it? CC @brijos @dblock 

Closes #1918

### CODE REVIEW

**Comment by sothawo** [comment]
That's the way it should go I think. Two big things to do:

### Configuration

The current code for Elasticsearch uses the same configuration class `org.springframework.data.elasticsearch.client.ClientConfiguration` for both the old code using RHLC (and the custom reactive code) and the new code using the Elasticsearch client (imperative and reactive); so this class is not in the `elc` or `erhlc` packages. This configuration class uses `org.elasticsearch.client.RestClientBuilder.HttpClientConfigCallback` that comes from the low level Elasticsearch client library, which is and always has been Apache 2 licensed. Does Opensearch use this same low level client or does it use an opensearch-renamed-copy of this as well? If the latter, then the opensearch integration needs a copy of the configuration class as well somewhere.

### Integration tests

I already prepared last year the integration of Opensearch into the integration test setup insofar as there is a property in `pom.xml` (`mvn.integration-test-opensearch.goal`) which should be set to _integration-test_ to add OS integration tests. In addition to this property, 

* the integration tests using Elasticsearch (`*ELCIntegrationTests` and `*ERHLCIntegrationTests`) should get an additional annotation `@EnabeldIfSystemProperty(named="sde.integration-test.environment" matches="elasticsearch")`)
* every test should get a new copy `*ORHLCIntegrationTests` which uses the new Opensearch configuration and is annotated with `@EnabeldIfSystemProperty(named="sde.integration-test.environment" matches="opensearch")`

And _src/test/resources/testcontainers-opensearch.properties_ needs to be set up to define and configure the correct opensearch docker image to be used by TestContainers.

I did not now try this out now, just writing what I assume should be working.

**Comment by reta** [comment]
@sothawo thanks a lot for the feedback and insights.

> This configuration class uses org.elasticsearch.client.RestClientBuilder.HttpClientConfigCallback that comes from the low level Elasticsearch client library, which is and always has been Apache 2 licensed. Does Opensearch use this same low level client or does it use an opensearch-renamed-copy of this as well?

Yes, it uses the same underlying `HttpAsyncClientBuilder` but the callback is repackaged under `org.opensearch.client.RestClientBuilder.HttpClientConfigCallback`, I will take care of that, thank you.

> I already prepared last year the integration of Opensearch into the integration test setup insofar as there is a property in pom.xml (mvn.integration-test-opensearch.goal) which should be set to integration-test to add OS integration tests. 

Yes, saw that as well, will continue the work you have started and make the changes to ensure both test suites are working, thank you.

File: src/main/java/org/springframework/data/elasticsearch/client/InetSocketAddressParser.java [hunk 1]
```diff
@@ -24,9 +24,10 @@
  * Utility to parse endpoints in {@code host:port} format into {@link java.net.InetSocketAddress}.
  *
  * @author Mark Paluch
+ * @author Andriy Redko
  * @since 3.2
  */
-class InetSocketAddressParser {
+public class InetSocketAddressParser {
```
Comment (reta) [inline src/main/java/org/springframework/data/elasticsearch/client/InetSocketAddressParser.java:30]: Making it public to be used by `org.springframework.data.elasticsearch.client.orhlc.DefaultClientConfiguration`

**Comment by reta** [comment]
@sothawo I believe `Opesearch`'s `RestHighLevelClient` support is complete (beside the documentation, working on it now), may I ask you to take a look please? Thank you.

**Comment by sothawo** [comment]
Hi, I will have a look these days. 
For your info: I will be in a call probably next week with a Product Manager for OpenSearch from Amazon and with the Spring Data project lead from VMware, talking about where to host and maintain the OS integration and how it might be handled in the future. So as I said, I will a a thorough look at your work, but will wait with a merge until after that call.

**Comment by reta** [comment]
> So as I said, I will a a thorough look at your work, but will wait with a merge until after that call.

Sure, thanks a lot!

**Comment by ralberts** [comment]
Excited to see this progressing!

**Comment by sothawo** [comment]
looks alright for me as a start on the new repository. 

I just changed the `InetSocketAddressParser` visiblity in _main_ (#2227) so that when you rebase on _main_ all changes from your side would then be in the new `orhlc` package or in the test setup.

**Comment by Mr-Smith-13** [comment]
I was wondering if this PR is merged soon because we would highly appreciate to be able to use the OpenSearch client to connect to our OpenSearch cluster.
Is there already a planned date / is it clear in which version this change wuill be included?

**Comment by sothawo** [comment]
This PR won't be merged. Please read the comment on the issue in https://github.com/spring-projects/spring-data-elasticsearch/issues/1918#issuecomment-1182798335 and https://github.com/opensearch-project/opensearch-clients/issues/28.

After the repository is created by Opensearch the code from this PR will be migrated over there

**Comment by reta** [comment]
Closing in favor of https://github.com/opensearch-project/spring-data-opensearch/pull/1 as per discussions with Spring Data team

**Pull request closed without merging**